### PR TITLE
spacetime publish: Add `--wasm-file` flag

### PR DESCRIPTION
**Pull request by bfops**
# Summary of Changes
These change(s) include:
- Interface changes
    - [ ] API/ABI changes
    - [ ] Changes to the Rust library interface
    - [x] Command-Line interface (CLI)
- [ ] Bugfixes
- [ ] Performance improvements
- [ ] Refactoring
- [ ] Other (briefly describe)

# Description of Changes
This adds a `spacetime publish --wasm-file` flag, which allows the user to bypass building entirely.

This functionality already existed under `--project-path`, but was not documented or discoverable.

This PR also makes the `--project-path` and `--wasm-file` flags consistent in behavior between `spacetime publish`, `spacetime generate`, and `spacetime build`.

# Breaking changes
No breaking changes. This adds a new flag that did not previously exist.

We deprecated the undocumented usage of the old flag.

# Expected complexity level and risk
1?

# Testing
I used [cram](https://bitheap.org/cram/) to test this change:

```
Normal publishing still works:
  $ spacetime publish -s localhost test-publish-norm -c -S
  info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
      Finished release [optimized] target(s) in 0.02s
  Optimising module with wasm-opt...
  Could not find wasm-opt to optimise the module.
  For best performance install wasm-opt from https://github.com/WebAssembly/binaryen/releases.
  Continuing with unoptimised module.
  Uploading to localhost => http://127.0.0.1:3000
  Updated database with domain: test-publish-norm, address: 93dda09db9a56d8fa6c024d843e805d8

We get a warning if we do the obscure thing:
  $ spacetime publish -s localhost test-publish-wasm -c -S --project-path target/wasm32-unknown-unknown/release/spacetime_module.wasm
  Note: Using --project-path to provide a wasm file is deprecated, and will be
  removed in a future release. Please use --wasm-file instead.
  Uploading to localhost => http://127.0.0.1:3000
  Updated database with domain: test-publish-wasm, address: 5fc9c102c694a8bd7a96e3588bd0f760

Publishing a wasm file works as intended:
  $ spacetime publish -s localhost test-publish-wasm -c -S --wasm-file target/wasm32-unknown-unknown/release/spacetime_module.wasm
  Skipping build. Instead we are publishing target/wasm32-unknown-unknown/release/spacetime_module.wasm
  Uploading to localhost => http://127.0.0.1:3000
  Updated database with domain: test-publish-wasm, address: 5fc9c102c694a8bd7a96e3588bd0f760

...and it definitely isn't rebuilding:
  $ rm target/wasm32-unknown-unknown/release/spacetime_module.wasm
  $ spacetime publish -s localhost test-publish-wasm -c -S --wasm-file target/wasm32-unknown-unknown/release/spacetime_module.wasm
  Skipping build. Instead we are publishing target/wasm32-unknown-unknown/release/spacetime_module.wasm
  Error: No such file or directory (os error 2)
  [1]
```

### CODE REVIEW

File: crates/cli/src/subcommands/publish.rs [hunk 1]
```diff
@@ -137,10 +145,16 @@ pub async fn exec(mut config: Config, args: &ArgMatches) -> Result<(), anyhow::E
         query_params.push(("trace_log", "true"));
     }
 
-    let path_to_wasm = if !path_to_project.is_dir() && path_to_project.extension().map_or(false, |ext| ext == "wasm") {
-        path_to_project.clone()
+    let path_to_wasm;
+    if !path_to_project.is_dir() && path_to_project.extension().map_or(false, |ext| ext == "wasm") {
+        println!("Note: Using --project-path to provide a wasm file is deprecated, and will be");
+        println!("removed in a future release. Please use --precompiled-wasm instead.");
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:151]: I think this `--precompiled-wasm` flag is wrong?

**Review by jdetter — CHANGES_REQUESTED** [review]

File: crates/cli/src/subcommands/publish.rs [hunk 2]
```diff
@@ -137,10 +145,16 @@ pub async fn exec(mut config: Config, args: &ArgMatches) -> Result<(), anyhow::E
         query_params.push(("trace_log", "true"));
     }
 
-    let path_to_wasm = if !path_to_project.is_dir() && path_to_project.extension().map_or(false, |ext| ext == "wasm") {
-        path_to_project.clone()
+    let path_to_wasm;
+    if !path_to_project.is_dir() && path_to_project.extension().map_or(false, |ext| ext == "wasm") {
+        println!("Note: Using --project-path to provide a wasm file is deprecated, and will be");
+        println!("removed in a future release. Please use --wasm-file instead.");
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:151]: One last request, could we make a new PR which is breaking which removes this (mark the PR with release 0.9 and ABI/API break)

File: crates/cli/src/subcommands/publish.rs [hunk 3]
```diff
@@ -39,6 +39,13 @@ pub fn cli() -> clap::Command {
                 .short('p')
                 .help("The system path (absolute or relative) to the module project")
         )
+        .arg(
+            Arg::new("wasm_file")
+                .value_parser(clap::value_parser!(PathBuf))
+                .long("wasm-file")
+                .short('w')
+                .help("The system path (absolute or relative) to the wasm file we should publish, instead of building the project."),
```
Comment (jdetter) [inline crates/cli/src/subcommands/publish.rs:47]: As ingvar stated in Discord, the `spacetime generate` command also has this wonkyness of accepting `--project-path`, could we update it there as well? Thanks Zeke 🙏

File: crates/cli/src/subcommands/publish.rs (hunk 2 above)
Comment (bfops) [reply crates/cli/src/subcommands/publish.rs:151]: WIP but PR started here https://github.com/clockworklabs/SpacetimeDB/pull/915

File: crates/cli/src/subcommands/publish.rs (hunk 3 above)
Comment (bfops) [reply crates/cli/src/subcommands/publish.rs:47]: Thank you! Brought `spacetime publish` in line with `spacetime generate` and `spacetime build`.

File: crates/cli/src/subcommands/build.rs [hunk 4]
```diff
@@ -7,10 +7,10 @@ pub fn cli() -> clap::Command {
     clap::Command::new("build")
         .about("Builds a spacetime module.")
         .arg(
-            Arg::new("project-path")
-                .default_value(".")
+            Arg::new("project_path")
                 .value_parser(clap::value_parser!(PathBuf))
-                .help("The path of the project that you would like to build."),
+                .default_value(".")
+                .help("The system path (absolute or relative) to the project you would like to build")
```
Comment (bfops) [inline crates/cli/src/subcommands/build.rs:13]: tried to use more consistent text, arg ordering, and naming between subcommands where possible

File: crates/cli/src/subcommands/publish.rs [hunk 5]
```diff
@@ -32,13 +32,21 @@ pub fn cli() -> clap::Command {
                 .help("When publishing a new module to an existing address, also delete all tables associated with the database"),
         )
         .arg(
-            Arg::new("path_to_project")
+            Arg::new("project_path")
```
Comment (bfops) [inline crates/cli/src/subcommands/publish.rs:35]: tried to use more consistent naming between subcommands where possible

File: crates/cli/src/subcommands/generate/mod.rs [hunk 6]
```diff
@@ -107,20 +106,15 @@ pub fn exec(args: &clap::ArgMatches) -> anyhow::Result<()> {
...
-        },
+    let wasm_file = if !project_path.is_dir() && project_path.extension().map_or(false, |ext| ext == "wasm") {
+        println!("Note: Using --project-path to provide a wasm file is deprecated, and will be");
+        println!("removed in a future release. Please use --wasm-file instead.");
+        project_path.clone()
+    } else if let Some(path) = wasm_file {
+        println!("Skipping build. Instead we are inspecting {}", path.display());
+        path.clone()
+    } else {
+        crate::tasks::build(project_path, skip_clippy, build_debug)?
```
Comment (bfops) [inline crates/cli/src/subcommands/generate/mod.rs:117]: this structure, param handling, and error behavior is more in line with what `publish` does (see below)

**Review by jdetter — APPROVED** [review]
LGTM 👍

File: crates/cli/src/subcommands/generate/mod.rs [hunk 7]
```diff
@@ -34,11 +34,10 @@ pub fn cli() -> clap::Command {
         .arg(
             Arg::new("project_path")
                 .value_parser(clap::value_parser!(PathBuf))
+                .default_value(".")
                 .long("project-path")
                 .short('p')
-                .default_value(".")
-                .conflicts_with("wasm_file")
```
Comment (bfops) [inline crates/cli/src/subcommands/generate/mod.rs:40]: `project_path` already conflicts with `wasm_file`

**Pull request merged**